src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -390,25 +390,45 @@ entails SolverOptions{..} constraint context hints =
         subclassDictionaryValue dict className index =
           App (Accessor (mkString (superclassName className index)) dict) valUndefined
 
-    solveCoercible :: Environment -> [SourceType] -> [SourceType] -> m (Maybe [TypeClassDict])
-    solveCoercible env kinds [a, b] = runMaybeT $ do
+    solveCoercible :: Environment -> InstanceContext -> [SourceType] -> [SourceType] -> m (Maybe [TypeClassDict])
+    solveCoercible env ctx kinds [a, b] = runMaybeT $ do
       let kindOf = lift . (sequence . (id &&& elaborateKind)) <=< replaceAllTypeSynonyms
       (a', kind) <- kindOf a
       (b', kind') <- kindOf b
       lift $ unifyKinds kind kind'
-      -- Solving terminates when the two arguments are the same. Since we
-      -- currently don't support higher-rank arguments in instance heads, term
-      -- equality is a sufficient notion of "the same".
-      if a' == b'
+      -- Solving terminates when the two arguments are the same or if a
+      -- dictionary is already in scope. Since we currently don't support
+      -- higher-rank arguments in instance heads, term equality is a sufficient
+      -- notion of "the same".
+      let coercibleDictsInScope = findDicts ctx C.Coercible Nothing
+      if a' == b' || any ((||) <$> isCoercibleDictInScope a' b' <*> isSymmetricCoercibleDictInScope a' b') coercibleDictsInScope
         then pure [TypeClassDictionaryInScope [] 0 EmptyClassInstance [] C.Coercible [] kinds [a, b] Nothing]
         else do
           -- When solving must reduce and recurse, it doesn't matter whether we
           -- reduce the first or second argument -- if the constraint is
           -- solvable, either path will yield the same outcome. Consequently we
           -- just try the first argument first and the second argument second.
-          ws <- (MaybeT $ coercibleWanteds env a' b') <|> (MaybeT $ coercibleWanteds env b' a')
+          -- If both path fail to yield subgoals we try to make some progress
+          -- regardless by transitivity.
+          subst <- lift $ gets checkSubstitution
+          let k = substituteType subst kind
+          ws <- MaybeT (coercibleWanteds env a' b')
+            <|> MaybeT (coercibleWanteds env b' a')
+            <|> guarded (not . null) (mapMaybe (srcCoercibleTransitiveConstraint k a' b') coercibleDictsInScope)
           pure [TypeClassDictionaryInScope [] 0 EmptyClassInstance [] C.Coercible [] kinds [a, b] (Just ws)]
-    solveCoercible _ _ _ = pure Nothing
+    solveCoercible _ _ _ _ = pure Nothing
+
+    isCoercibleDictInScope a b TypeClassDictionaryInScope{..} = tcdInstanceTypes == [a, b]
+
+    isSymmetricCoercibleDictInScope a b TypeClassDictionaryInScope{..} = tcdInstanceTypes == [b, a]
+
+    srcCoercibleTransitiveConstraint k a b TypeClassDictionaryInScope{..}
+      | [a', b'] <- tcdInstanceTypes =
```

Comment:
I'm a little bit unsure about this approach. Could it lead to a loop? Maybe we should be following section 5 of the paper a bit more closely here? Transitivity is hard because it's not always going to be clear what type to pick for `b` to get from `a` to `c`, so I think we should expect the solver to fail to solve Coercible in at least some of the harder cases which require use of transitivity rule. Perhaps we should just do symmetry on its own first?

Reply:
Yeah, along those lines, here's a case for which the patch currently produces a PossiblyInfiniteInstance:
```purs
test :: forall a b c d. Coercible a c => Coercible a d => Coercible d b => Proxy c -> Proxy d -> a -> b
test _ _ = coerce
```
[The equivalent Haskell](https://tio.run/##VU7LCsIwELznK@aYQPUDCvGiFw@ifyB5rDSYmpK2qF8fk1ShsrAzs@zOTqfGO3mfUh/s7Akn5R54dhSJMdcPIU44qElt94GiIfCKTntqYOpI/K1dYni9wSsIxm4hoG1XV1AwDdbSrqWFFpA7LDYGmx@1harSdDW95pLfBIz1JXT@czyDi0VJDHPMeUVKHw "Haskell – Try It Online") works, though I know nothing about how what this patch does differs from what GHC is doing.

If a temporary halfway solution is a good idea, though, wouldn't it be better to have the simpler transitive cases covered with some weirder ones producing PossiblyInfiniteInstance errors then to have all transitive cases produce NoInstanceFound errors?

Maybe a counterargument to that is that this transitive search can produce PossiblyInfiniteInstances where NoInstanceFound would be correct and clearer, in cases like `forall a. Coercible a String => a -> Int`. Does that matter much? I think maybe it does.

Reply:
Yeah, I think producing PossiblyInfiniteInstance when we should be producing NoInstanceFound ought to be considered a bug. I think temporary halfway solutions are best avoided in the solver in general, as fixing a temporary halfway solution to work properly might mean that things which used to type check no longer do, and pushing out a change like that is usually going to ruin a few people’s days.

Reply:
Oh that’s no good, I‘ve undone transitivity. Thank you for noticing!

Reply:
For the record, I‘ve read the section **5 Type inference with Coercible constraints** of the paper and I‘m trying to rewrite the solver of `Coercible` constraints to support much richer interactions of givens and wanteds, as explained in the paper.